### PR TITLE
Support dual stack ipset

### LINKED ISSUE #395: Multiple ipsets supoort for one domain
**需求应用场景**
I wanna use [ss-rules](https://github.com/openwrt/packages/blob/openwrt-19.07/net/shadowsocks-libev/files/ss-rules) for smart forward of shadowsocks. It creates some ipsets for this purpose, among which `ss_rules_dst_forward` and `ss_rules6_dst forward` for ipv4 and ipv6 respectively could be utilized by smartdns. So I wrote rules like:
```
nameserver /google.com/oversea # use oversea group for this domain
ipset /google.com/ss_rules_dst_forward # for ipv4
ipset /google.com/ss_rules6_dst_forward # for ipv4
```
However it didn't work as expected. Only `ss_rules6_dst_forward` took effect and I can only had ipv6 traffic forwarded.

**建议的方案**
The resolved result can be added to all ipsets associated with this domain.
I scan the code. Maybe changing `domain_rule.rules[DOMAIN_RULE_IPSET]->ipset_name` to an array is helpful.

**设备信息**
1. 设备信息（CPU，厂家）  
x86 soft router
2. 固件信息
Openwrt 19.07. Latest ver of smartdns compiled by myself.

### PR DESCRIPTION
**Pull request by yujincheng08**
Close #395.

### CODE REVIEW

**Comment by zonyitoo** [comment]
@pymumu Please consider merging this PR

File: src/dns_server.c [hunk 1]
```diff
@@ -691,32 +691,56 @@ static int _dns_setup_ipset(struct dns_request *request)
 
 	/* check ipset rule */
 	rule_flags = request->domain_rule.rules[DOMAIN_RULE_FLAGS];
-	if (rule_flags) {
-		if (rule_flags->flags & DOMAIN_FLAG_IPSET_IGNORE) {
-			return 0;
-		}
+	if (!rule_flags || (rule_flags->flags & DOMAIN_FLAG_IPSET_IGN) != 0) {
```
Comment (zonyitoo) [inline src/dns_server.c:694]: It should be:

```c
if (!rule_flags || (rule_flags->flags & DOMAIN_FLAG_IPSET_IGN) == 0)
```

File: src/dns_server.c [hunk 2]
```diff
@@ -691,32 +691,56 @@ static int _dns_setup_ipset(struct dns_request *request)
...
+		ipset_rule = request->domain_rule.rules[DOMAIN_RULE_IPSET];
+	}
+	if (!rule_flags || (rule_flags->flags & DOMAIN_FLAG_IPSET_IPV4_IGN) != 0) {
```
Comment (zonyitoo) [inline src/dns_server.c:697]: Same as above.

File: src/dns_server.c [hunk 3]
```diff
@@ -691,32 +691,56 @@ static int _dns_setup_ipset(struct dns_request *request)
...
+		ipset_rule_v4 = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV4];
+	}
+	if (!rule_flags || (rule_flags->flags & DOMAIN_FLAG_IPSET_IPV6_IGN) != 0) {
```
Comment (zonyitoo) [inline src/dns_server.c:700]: Same as above.

File: src/dns_server.c [hunk 4]
```diff
@@ -1815,13 +1838,18 @@ static int _dns_server_setup_ipset_packet(struct dns_request *request, struct dn
...
-		if (rule_flags->flags & DOMAIN_FLAG_IPSET_IGNORE) {
-			return 0;
+		if ((rule_flags->flags & DOMAIN_FLAG_IPSET_IGN) == 0) {
+			ipset_rule = request->domain_rule.rules[DOMAIN_RULE_IPSET];
+		}
+		if ((rule_flags->flags & DOMAIN_FLAG_IPSET_IPV4_IGN) == 0) {
+			ipset_rule_v4 = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV4];
+		}
+		if ((rule_flags->flags & DOMAIN_FLAG_IPSET_IPV6_IGN) == 0) {
+			ipset_rule_v6 = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV6];
```
Comment (zonyitoo) [inline src/dns_server.c:1848]: Should be:

```c
                if (!rule_flags || (rule_flags->flags & DOMAIN_FLAG_IPSET_IGN) == 0) {
			ipset_rule = request->domain_rule.rules[DOMAIN_RULE_IPSET];
		}
		if (!rule_flags || (rule_flags->flags & DOMAIN_FLAG_IPSET_IPV4_IGN) == 0) {
			ipset_rule_v4 = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV4];
		}
		if (!rule_flags || (rule_flags->flags & DOMAIN_FLAG_IPSET_IPV6_IGN) == 0) {
			ipset_rule_v6 = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV6];
		}
```

**Comment by pymumu** [comment]
is all work done?

**Comment by yujincheng08** [comment]
> is all work done?

AFAIK, yes.

**Pull request merged**